### PR TITLE
Javascript Error Tracking

### PR DESCRIPTION
**Pull request by MikeRogers0**
This adds the `js_api_key` configuration option, if it's present it'll inject the JS code to a HTML page.

Some key changes:
 - I moved the folder `specs` to `spec` to be more inline with other projects I've worked on.
 - I added a dummy app for full integration tests
 - I changed the RSpec syntax to use `expect( something ).to eq( something else )` as it's closer to their sample docs.
 - I upgraded a few of the gems to be the latest.
 - I added SimpleCov to track coverage of RSpec

This is for the issue #132

### CODE REVIEW

**Comment by UberMouse** [comment]
Thanks for this Mike. I'll take a look when I get time, either this week or the next.

**Comment by MikeRogers0** [comment]
@UberMouse Thank you for reviewing :) It's a big change, so take your time & enjoy it.

File: spec/features/javascript_spec.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+feature 'JavaScript Tracking', feature: true do
+  before { Raygun.configuration.js_api_key = nil }
+  after { Raygun.configuration.js_api_key = nil }
+
+  it "Includes the Raygun Javascript Middleware" do
+    expect(Rails.application.config.middleware).to include(Raygun::Middleware::JavascriptExceptionTracking)
+  end
+
+  it "No JS snippet is injected" do
```
Comment (UberMouse) [inline spec/features/javascript_spec.rb]: I would wrap this in a context 'Without JS Api key'. I'm also a fan of the `it` being included in the test name so would rephrase to something more like `it 'does not inject JS snippet'`

File: spec/features/javascript_spec.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    visit root_path
+
+    expect(page.html).to_not include("<script type=\"text/javascript\">!function(a,b,c,d,e,f,g,h){a.RaygunObject=e,a[e]=a[e]||function(){(a[e].o=a[e].o||[]).push(arguments)},f=b.createElement(c),g=b.getElementsByTagName(c)[0],f.async=1,f.src=d,g.parentNode.insertBefore(f,g),h=a.onerror,a.onerror=function(b,c,d,f,g){h&&h(b,c,d,f,g),g||(g=new Error(b)),a[e].q=a[e].q||[],a[e].q.push({e:g})}}(window,document,\"script\",\"//cdn.raygun.io/raygun4js/raygun.min.js\",\"rg4js\");</script></head>")
```
Comment (UberMouse) [inline spec/features/javascript_spec.rb]: Probably better to assert something simpler here, maybe just checking for the presence of `raygun.min.js`?

File: spec/features/javascript_spec.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    expect(page.html).to_not include("<script type=\"text/javascript\">rg4js('apiKey', '');rg4js('enableCrashReporting', true);</script></body>")
```
Comment (UberMouse) [inline spec/features/javascript_spec.rb]: Same for here, if you update this snippet this test will break. Probably just want to check for `rg4js`

File: spec/features/javascript_spec.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+  end
+
+  context 'With a JS API Key' do
+    before { Raygun.configuration.js_api_key = 'Sample key' }
+
+    it "JS snippet is injected requesting a HTML file" do
```
Comment (UberMouse) [inline spec/features/javascript_spec.rb]: Same changes as above for these two

File: lib/raygun/javascript_tracker.rb [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        "rg4js('enableCrashReporting', true);",
+        user_context,
+        '</script>'
+      ].join('').html_safe
+    end
+
+    private
+
+    def user_context
+      # TODO, Pull in user data from somewhere.
```
Comment (UberMouse) [inline lib/raygun/javascript_tracker.rb]: Yes, you should be able to just pull the hash out of the ENV and pass it in as `rg4js('setUser', ...)`. I'm not exactly sure how you can convert it to a JS object but shouldn't be too hard. The hash fields are already mapped to the same fields Raygun4JS expects.

File: lib/raygun/javascript_tracker.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
+# Client for injecting JavaScript code for tracking front end exceptions
```
Comment (UberMouse) [inline lib/raygun/javascript_tracker.rb]: This looks fine. The only other config option I would bother including for now would be the version if it's set

**Comment by MikeRogers0** [comment]
Awesome thank you for the review, I'll aim to get on it possibly next weekend :)

**Comment by MikeRogers0** [comment]
Eak my weekend got away from me a bit - don't worry, I've not forgotten about this :)

File: lib/raygun/javascript_tracker.rb (hunk 5 above)
Comment (MikeRogers0) [reply lib/raygun/javascript_tracker.rb]: I decided to leave this for another PR, I think users who are interested in the setting their user context should be ok to read the docs (Plus I'm keen to get this in ;) )

**Comment by MikeRogers0** [comment]
Thanks for your patience @UberMouse 👍

I got a bit of time this evening to do the updated you requested (and respond to the other comments). Any other changes (I might have some time over NYE)?

File: lib/raygun/javascript_tracker.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    end
+
+    private
+
+    def js_api_key
+      @js_api_key ||= Raygun.configuration.js_api_key
+    end
+
+    def js_api_version
+      @js_api_version ||= Raygun.configuration.js_api_version || '1.14.0'
```
Comment (UberMouse) [inline lib/raygun/javascript_tracker.rb]: This should default to `''` so that it always falls back to the latest version unless you specifically set it to another version. The url should be fine if it ends up like `//cdn.raygun.io/raygun4js//raygun.min.js`

Comment (MikeRogers0) [reply lib/raygun/javascript_tracker.rb]: @UberMouse Updated :D

**Comment by UberMouse** [comment]
Thanks for that Mike. I tested it and that URL format with double slashes in it doesn't work on S3 so I've just gone ahead and fixed that part up since it was my suggestion. I've opened a new PR and will merge that work in shortly over there.
